tests/property/ic_api/canister_version/test/tests.ts

```diff
@@ -0,0 +1,129 @@
+import {
+    defaultPropTestParams,
+    expect,
+    getCanisterActor,
+    it,
+    Test
+} from 'azle/test';
+import { execSync } from 'child_process';
+import fc from 'fast-check';
+
+import { _SERVICE as Actor } from './dfx_generated/canister/canister.did';
+
+export function getTests(): Test {
+    return () => {
+        it('gets the canister version from the canister', async () => {
+            await fc.assert(
+                fc.asyncProperty(
+                    fc.nat({
+                        max: 10
+                    }),
+                    async (timesToDeploy) => {
+                        // Stop and delete the canister before every run
+                        execSync(`dfx canister stop canister || true`, {
+                            stdio: 'inherit'
+                        });
+
+                        execSync(`dfx canister delete canister || true`, {
+                            stdio: 'inherit'
+                        });
+
+                        execSync(`dfx deploy canister`, {
+                            stdio: 'inherit'
+                        });
+
+                        const startingVer = 1n; // Canister starts at 0 when it's created but the deploy in pretest should increment it to be at 1 by the time it gets here
+
+                        const actor = await getCanisterActor<Actor>('canister');
+
+                        await checkInitVersion(actor, startingVer);
+                        await checkUpgradeVersion(actor, []);
+                        await checkInspectVersion(actor, startingVer);
+                        await checkQueryAndUpdateVersion(
+                            actor,
+                            startingVer + 1n
+                        );
+
+                        const versionsAtDeploy = Array.from(
+                            { length: timesToDeploy },
+                            (_, index) => BigInt(index * 2) + startingVer
+                        );
+
+                        for (const version of versionsAtDeploy) {
+                            execSync(`dfx deploy canister --upgrade-unchanged`);
+
+                            await checkUpgradeVersion(actor, [version]);
+                            await checkQueryAndUpdateVersion(actor, version);
+                        }
+                    }
+                ),
+                defaultPropTestParams
+            );
+        });
+    };
+}
+
+async function checkInitVersion(
+    actor: Actor,
+    expectedVer: bigint
+): Promise<void> {
+    const initCanisterVersion = await actor.getInitCanisterVersion();
+
+    expect(initCanisterVersion).toHaveLength(1);
+    expect(initCanisterVersion[0]).toBeGreaterThanOrEqual(expectedVer);
+}
+
+async function checkUpgradeVersion(
+    actor: Actor,
+    expectedVersion: [bigint] | []
```

Comment:
Why does the expected version look like an option here?

Reply:
because before an upgrade happens the pre upgrade will not have been called so we will be expecting None after the first init and Some after post_upgrade